api/hasura/actions/vouch.ts

```diff
@@ -0,0 +1,151 @@
+    // vouch and build user if needed
+    const updatedNominee = await vouch(nomineeId, voucher);
+
+    return okResponse(res, { id: updatedNominee.id });
+  } catch (e: any) {
+    return errorResponse(res, e);
+  }
+}
+
+async function validate(nomineeId: number, voucherProfileId: number) {
+  // Get the nominee
+  const nominee = await getNominee(nomineeId);
+
+  // make sure circle exists
+  if (!nominee.circle) {
+    throw new NotFoundError('circle not found');
+  }
+
+  // Check if voucher exists in the same circle as the nominee
+  // TODO: this uses assert for error handling
+  const voucher = await getUserFromProfileId(
+    voucherProfileId,
+    nominee.circle_id
+  );
+
+  // If circle only allows giver to vouch, make sure voucher is a giver
+  if (nominee.circle.only_giver_vouch && voucher.non_giver) {
+    throw new ForbiddenError(
+      "voucher is a 'non-giver' so is not allowed to vouch"
+    );
+  }
+
+  // make sure the nomination period hasn't ended
+  if (nominee.ended) {
+    throw new BadRequestError('nomination has already ended for this nominee');
+  }
+
+  // TODO: could this be handled by a unique index in the vouches table?
+  // Check if voucher already has an existing vouch for the nominee
+  if (nominee.nominated_by_user_id === voucher.id) {
+    throw new ForbiddenError(
+      "voucher nominated this nominee so can't additionally vouch"
+    );
+  }
+
+  const { vouches } = await gql.getExistingVouch(nomineeId, voucher.id);
+
+  if (vouches.pop()) {
+    throw new ForbiddenError('voucher has already vouched for this nominee');
+  }
+
+  return {
+    voucher,
+  };
+}
+
+async function vouch(nomineeId: number, voucher: Voucher) {
+  // vouch for the nominee
+
+  // this inserts the vouch and also fetches the nominee with updated vouch count
+  const insert_vouches = await gql.insertVouch(nomineeId, voucher.id);
+  if (!insert_vouches?.nominee) {
+    throw new InternalServerError('unable to add vouch');
+  }
+
+  // the vouching is announced with an event trigger
+
+  const nominee = insert_vouches.nominee;
+
+  // if there are enough nominations, go ahead and add the user to the circle
+  const nomCount = nominee.nominations_aggregate.aggregate?.count || 0;
+  if (nominee.vouches_required - 1 <= nomCount) {
+    return await convertNomineeToUser(nominee);
+  }
+
+  return nominee;
+}
+
+async function convertNomineeToUser(nominee: Nominee) {
+  // Get the nominee into the user table
+  let userId = nominee.user_id;
+  if (!userId) {
+    const addedUser = await gql.insertUser(
+      nominee.address,
+      nominee.name,
+      nominee.circle_id
+    );
+    if (!addedUser) {
+      throw new InternalServerError('unable to add user');
+    }
+    userId = addedUser.id;
+  }
+
+  // The profile is automatically created by the createProfile event trigger, if needed
+
+  // attach the user id to the nominee, and mark the nomination ended
+  const updatedNominee = await gql.updateNomineeUser(nominee.id, userId);
```

Comment:
It's too bad we can't make this atomic. Maybe we could add a FK relationship in the other direction (a column in `users` that backlinks to a `nominee` row) instead of the current relationship. Then a nominee's success would be determined by a join in the gql query and we'd be able to do this atomically.

Reply:
Can we TODO this and add a ticket for later? This is a general concern I have for TX stuff that mixes DB+bizlogic

Reply:
I'm also unsure if its true that user ALWAYS has nominee. anyway i think this is a problem for later?

Reply:
i don't think `user_id` in `nominees` table is used currently anywhere , i added it just incase someone wanted to do analytics with this so it might not be that critical

Yes a user can be added directly by admin bypassing the vouching process so it wouldn't be linked to any nominee